### PR TITLE
Symplectic C1

### PR DESCRIPTION
**Pull request by TristanPfersdorff**
Add the necessary functions for the C1 subgroups in the symplectic case. The codecov bot will be happy once we add the symplectic main function

### General
- [ ] All functions have unit tests

### Functions constructing generators of maximal subgroups
- [ ] The code which computes and then stores the sizes is correct. To do this one confers to the tables referenced by the code. Tests for the group size should use `RECOG.TestGroup` from the recog package.
- [ ] The test suite checks whether all constructed generators are sensible. That is it tests the generators'
  - [ ] determinants (and if applicable also the spinor norms are correct), and
  - [ ] compatibility with the correct form,
  - [ ] `DefaultFieldOfMatrixGroup` returns the correct field.

### Functions assembling the list of all maximal subgroups of a certain group
- [ ] The code agrees with the tables in section 8.2 of the book "The Maximal Subgroups of the Low-dimensional Finite Classical Groups".

The reviewer doesn't need to compare our results to magma's results. That's the job of the person implementing the code.

### CODE REVIEW

File: gap/ReducibleMatrixGroups.gi [hunk 1]
```diff
@@ -358,3 +358,104 @@ function(d, q, k)
...
+    if k >= d / 2 then
+        ErrorNoReturn("<k> must be less than <d> / 2 but <k> = ", k, 
+        " and <d> = ", d);
+    fi;
+
+    field := GF(q);
+    gens := [];
+    twok := 2 * k;
+
+    # These generators are bidagonal block matrices
```
Comment (fingolfin) [inline gap/ReducibleMatrixGroups.gi:439]:
```suggestion
    # These generators are diagonal block matrices
```
Or perhaps you meant something else, but I am not sure what?

File: gap/ReducibleMatrixGroups.gi [hunk 2]
```diff
@@ -358,3 +358,104 @@ function(d, q, k)
...
 end);
+
+
+# Construction as in Proposition 4.3 of [HR05]
+BindGlobal("SpStabilizerOfIsotropicSubspace",
+function(d, q, k)
+    local field, gens, I, J, GLgen, Xi, Spgen, Yi;
+
+    if IsOddInt(d) then
+        ErrorNoReturn("<d> must even but ", d, " was given.");
```
Comment (fingolfin) [inline gap/ReducibleMatrixGroups.gi:369]: This line is not covered. Of course, it's an error, and you are careful to not trigger errors in your tests.

That said, in general I've come to find it useful to explicitly also add tests for functions with incorrect arguments, to verify that error handling is in place and working. As a bonus side effect, this also allows to get code coverage (closer) to 100%. It also helps to find things like formatting mistakes in error messages.

Note a requirement, thought, just a suggestion.

File: gap/ReducibleMatrixGroups.gi (hunk 1 above)
Comment (TristanPfersdorff) [reply gap/ReducibleMatrixGroups.gi:439]: It's supposed to be block matrices with blocks on both diagonals and zeros elsewhere, but I'm not sure if "bidiagonal" is common terminology, so I'll clarify the comment.

**Comment by TristanPfersdorff** [comment]
I've implemented all suggested changes including clarifying some comments and rebasing. The errors (e.g. for odd dimension) are now also checked.

File: gap/ClassicalMaximals.gi [hunk 3]
```diff
@@ -827,3 +827,17 @@ function(n, q, classes...)
 
     return maximalSubgroups;
 end);
+
+
+BindGlobal("C1SubgroupsSymplecticGroupGeneric",
+function(n, q)
+    local listOfks, result;
+    listOfks := [1..QuoInt(n, 2) - 1];
```
Comment (fingolfin) [inline gap/ClassicalMaximals.gi:835]: This is not covered, but my understanding this is because this function will only be called in a future PR, correct?

Comment (TristanPfersdorff) [reply gap/ClassicalMaximals.gi:835]: Correct. Theses PRs are only for the individual Classes, once they are all done we will call the generic functions from a function along the lines of `MaximalSubgroupClassRepsSymplecticGroup`. The generic functions will only be called from there.

**Review by fingolfin — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**